eth/watchers/roundswatcher.go

```diff
@@ -0,0 +1,160 @@
+package watchers
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+)
+
+// RoundsWatcher is a type alias for a thread safe in-memory cache that watches on-chain events for new initialized rounds
+type RoundsWatcher struct {
+	mu                   sync.RWMutex
+	lastInitializedRound *big.Int
+	blockHashForRound    [32]byte
+
+	quit chan struct{}
+
+	watcher Watcher
+	lpEth   eth.LivepeerEthClient
+	dec     *EventDecoder
+}
+
+type Watcher interface {
+	Subscribe(sink chan<- []*blockwatch.Event) event.Subscription
+}
+
+// NewRoundsWatcher creates a new instance of RoundsWatcher and sets the initial cache through an RPC call to an ethereum node
+func NewRoundsWatcher(watcher Watcher, lpEth eth.LivepeerEthClient, addr ethcommon.Address) (*RoundsWatcher, error) {
+	dec, err := NewEventDecoder(addr, contracts.RoundsManagerABI)
+	if err != nil {
+		return nil, fmt.Errorf("error creating decoder: %v", err)
+	}
+
+	return &RoundsWatcher{
+		quit:    make(chan struct{}),
+		watcher: watcher,
+		lpEth:   lpEth,
+		dec:     dec,
+	}, nil
+}
+
+// LastInitializedRound gets the last initialized from cache
+func (rw *RoundsWatcher) LastInitializedRound() (*big.Int, error) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.lastInitializedRound, nil
+}
+
+// BlockHashForRound gets the blockhash for the last initialized round from cache
+func (rw *RoundsWatcher) BlockHashForRound(round *big.Int) ([32]byte, error) {
```

Comment:
I suggest that we change this to `LastInitializedBlockHash()` and omit any arguments so that we don't have to worry about argument checks. Then, separately we can change the `pm.RoundsManager` interface method (the interface was originally written so that `eth.LivepeerEthClient` could be passed in directly, but now we can pass in `RoundsWatcher` so I think it is reasonable to also update to a cleaner API).

Reply:
I have reworked both methods then to not return an error either, since we're just reading values and don't have to make external calls.

I addressed the `pm` package specific changes in a seperate commit (and fixup).